### PR TITLE
Added the ability to log the request and response

### PR DESCRIPTION
**Pull request by Minx-SigEp**
Added the ability to log the request and response in the AvaTax Rest Client. This provides consumers of the Rest Client with a way of seeing the raw request and response objects for debugging and auditing.

### CODE REVIEW

**Comment by ted-spence-avalara** [comment]
Hi Minx - This is a very highly requested feature and I'd love to get this incorporated into the SDK!  I notice that this PR has a lot of other changes too beyond the logging.  Can we keep each PR separate for the moment so we can review each one in turn?

I'd like to focus on getting this reviewed and vetted by our security team so we can get this into the next release (likely in June).

File: src/AvaTaxError.cs [hunk 1]
```diff
@@ -10,13 +10,19 @@ public class AvaTaxError : Exception
         /// <summary>
         /// The raw error message from the client
         /// </summary>
-        public ErrorResult error { get; set; }
+        public ErrorResult error { get; private set; }
```
Comment (ted-spence-avalara) [inline src/AvaTaxError.cs]: Let's keep this out of the logging PR to keep it focused on one single feature at a time.

File: src/AvaTaxApi.cs [hunk 2]
```diff
@@ -3394,9 +3394,9 @@ public FileResult BuildPointOfSaleDataFile(PointOfSaleDataRequestModel model)
         public String ChangePassword(PasswordChangeModel model)
         {
             var path = new AvaTaxPath("/api/v2/passwords");
-            return RestCallString("Put", path, model);
+            return ExecuteRestCall("Put", path, model, BodyAsString);
```
Comment (ted-spence-avalara) [inline src/AvaTaxApi.cs]: It might be simpler to just modify RestCallString to do logging on its own, wouldn't it?

File: src/AvaTaxError.cs (hunk 1 above)
Comment (Minx-SigEp) [inline src/AvaTaxError.cs]: Ok will update to no longer be in the Pull Request.

**Comment by Minx-SigEp** [comment]
@ted-spence-avalara I built the request and response logging on top of the deadlock/ async changes made in #20 as I felt it was important to include these changes, and the refactoring I did in that branch also allowed me to only have to implement the logging logic in two places. 

How do you want to deal with this?

**Comment by ted-spence-avalara** [comment]
Thanks for the update - let me look at this PR this afternoon and see how close we are to merging.  I want to update the tests on net20 and net45 as well.

**Comment by Minx-SigEp** [comment]
Cool. If there is a list of task you would like me to do so this can get merged let me know and I will work on it. Any thing that helps my company be on the supported SDK or as close to the supported SDK as possible is always great.

File: src/AvaTaxError.cs (hunk 1 above)
Comment (Minx-SigEp) [inline src/AvaTaxError.cs]: Removed this change as requested in commit 14c7df5ddcb87b2f8978c983720e6c86eb69bd84

**Comment by ted-spence-avalara** [comment]
We have now added a generic logging / API call capture mechanism to the API in this pull request: https://github.com/avadev/AvaTax-REST-V2-DotNet-SDK/pull/41

I'm closing this old PR - I hope the solution will help you!

**Pull request closed without merging**